tensorflow/core/kernels/substr_op.cc

```diff
@@ -0,0 +1,145 @@
+        auto output = output_tensor->flat<string>();
+
+        // Set output to be substrings of input strings
+        for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+          OP_REQUIRES(context, pos >= 0 && pos < input(i).size(),
+                      errors::InvalidArgument("pos ", pos, 
+                                              " out of range for string b'", 
+                                              input(i), "' at index ", i));
+          output(i) = input(i).substr(pos, len);
+        }
+      } else if (input_shape == pos_shape) {
+        // Perform Op element-wise
+        auto input = input_tensor.flat<string>();
+        auto pos = pos_tensor.flat<T>();
+        auto len = len_tensor.flat<T>();
+
+        // Allocate output
+        Tensor* output_tensor = nullptr;
+        OP_REQUIRES_OK(context,
+                       context->allocate_output("output", input_tensor.shape(),
+                                                &output_tensor));
+        auto output = output_tensor->flat<string>();
+
+        // Set output to be substrings of input strings
+        for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+          OP_REQUIRES(context, pos(i) >= 0 && pos(i) < input(i).size(),
+                      errors::InvalidArgument("pos ", pos(i), 
+                                              " out of range for string b'", 
+                                              input(i), "' at index ", i));
+          output(i) = input(i).substr(pos(i), len(i));
+        }
+      } else {  
+        // Attempt broadcasting this Op
+        context->SetStatus(errors::Unimplemented(
+                 "Substr broadcast is not supported yet."));
+
+        // BCast bcast(BCast::FromShape(input_shape), BCast::FromShape(pos_shape));
+        // OP_REQUIRES(context, bcast.IsValid(), 
+        //             errors::InvalidArgument("Incompatible shapes: ", 
+        //                                     input_shape.DebugString(), " vs. ",
+        //                                     pos_shape.DebugString()));
+        // TensorShape output_shape = BCast::ToShape(bcast.result_shape());
+        // int ndims = output_shape.dims();
+        // Tensor* output_tensor = nullptr;
+        // OP_REQUIRES_OK(context,
+        //                context->allocate_output("output", output_shape,
+        //                                         &output_tensor));
+        // switch (ndims) {
+        //   case 2: {
+        //     auto output = output_tensor->shaped<string, 2>(bcast.x_reshape());
+        //     auto input_reshaped = input_tensor.shaped<string, 2>(bcast.x_reshape());
+        //     auto pos_reshaped = pos_tensor.shaped<T, 2>(bcast.y_reshape());
+        //     auto len_reshaped = len_tensor.shaped<T, 2>(bcast.y_reshape());
+              
+        //     typename TTypes<string, 2>::Tensor input = input_reshaped.broadcast(BCast::ToIndexArray<2>(bcast.x_bcast()));
```

Comment:
Ah, good point.  Well, if @benoitsteiner doesn't get back to you soon, perhaps stick with the scalar implementation for now, since I suspect that's good enough for your usecase.

Perhaps:

1) Implement the shape spec using broadcastable shapes.  E.g., have the C++ shape spec use BroadcastBinaryOpShapeFn in common_shape_fns.h so that the "spec" allows broadcasting: it's probably sufficient to check that shape(pos) == shape(len) and then do BroadcastBinaryOpShapefn on input and pos.

2) In your kernel implementation, check that the pos/len are scalar, and if not, return an errors::Unimplemented, so it's clear that it *should* be implemented but is not yet.

Probably also mention that true ternary broadcasting might be necessary to make it efficient, since the looping you're doing even for the scalar approach isn't parallel.